File: megumi/src/main/java/com/erisu/cloud/megumi/pcr/basic/logic/NameLogic.java
package com.erisu.cloud.megumi.pcr.basic.logic;

import com.erisu.cloud.megumi.pcr.basic.mapper.PcrAvatarMapper;
import com.erisu.cloud.megumi.pcr.basic.pojo.PcrAvatar;
import com.erisu.cloud.megumi.pcr.basic.pojo.PcrPrincess;
import com.erisu.cloud.megumi.util.FileUtil;
import com.erisu.cloud.megumi.util.MessageUtil;
import net.mamoe.mirai.contact.Group;
import net.mamoe.mirai.contact.User;
import net.mamoe.mirai.message.data.*;
import net.mamoe.mirai.utils.ExternalResource;
import org.springframework.stereotype.Component;

import javax.annotation.Resource;
import java.nio.file.Path;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * @Description NAME相关
 * @Author alice
 * @Date 2021/6/22 9:47
 **/
@Component
public class NameLogic {
    @Resource
    private PcrAvatarMapper avatarMapper;

    public Message getAvatar(User sender, Group group, String princessId) throws Exception {
        PcrAvatar pcrAvatar = avatarMapper.searchMaxStarAvatar(princessId);
        Path imagePath = FileUtil.downloadHttpUrl(pcrAvatar.getAvatarUrl(), "image", null);
        if (imagePath != null) {
            List<String> nameList = PcrInitData.INSTANCE.getIdMap().get(pcrAvatar.getPrincessId());
            CompletableFuture<Image> future = MessageUtil.INSTANCE.generateImageAsync(group, imagePath.toFile(), true);
            return MessageUtil.INSTANCE.message(new At(sender.getId()),
                    future.get(), new PlainText(nameList.get(0)));
        } else {
            return null;
        }

    }
}

File: megumi/src/main/java/com/erisu/cloud/megumi/pcr/basic/logic/NameLogic.kt
package com.erisu.cloud.megumi.pcr.basic.logic

import com.erisu.cloud.megumi.pcr.basic.mapper.PcrAvatarMapper
import com.erisu.cloud.megumi.pcr.basic.pojo.PcrPrincess
import com.erisu.cloud.megumi.pcr.basic.util.CharacterField
import com.erisu.cloud.megumi.util.FileUtil
import com.erisu.cloud.megumi.util.StreamMessageUtil
import net.mamoe.mirai.contact.Group
import net.mamoe.mirai.contact.User
import net.mamoe.mirai.message.data.*
import org.springframework.stereotype.Component
import javax.annotation.Resource

/**
 *@Description NAME相关
 *@Author alice
 *@Date 2021/6/24 10:18
 **/
@Component
class NameLogic {

    @Resource
    private lateinit var avatarMapper: PcrAvatarMapper

    @Resource
    private lateinit var pcrInitData: PcrInitData

    @Resource
    private lateinit var gachaLogic: GachaLogic

    @Throws(Exception::class)
    suspend fun getAvatar(sender: User, group: Group, princessId: String): Message {
        val avatarImage = getAvatarImage(group, princessId)
        return if (avatarImage == null) {
            messageChainOf(
                At(sender.id),
                PlainText(pcrInitData.idMap[princessId]!![0])
            )
        } else {
            messageChainOf(
                At(sender.id),
                avatarImage,
                PlainText(pcrInitData.idMap[princessId]!![0])
            )
        }

    }


    @Throws(Exception::class)
    suspend fun getAvatarImage(group: Group, princessId: String): Message? {
        val (_, _, _, avatarUrl) = avatarMapper.searchMaxStarAvatar(princessId)
        val imageResponse = FileUtil.downloadHttpUrl(avatarUrl, "image", null, null) ?: return null
        if (imageResponse.code != 200) return null
        return StreamMessageUtil.generateImage(group, imageResponse.path!!.toFile(), true)
    }

    fun findCharacter(sender: User, rawMessage: String, group: Group, regexStr: String): Message {
        val regex = Regex(regexStr)
        val type = regex.find(rawMessage)!!.groupValues[1]
        val value = regex.find(rawMessage)!!.groupValues[2]
        val characterField = CharacterField.values().first { it.key == type }
        val properties = PcrPrincess::class.java.declaredFields
        val resultList: MutableList<Int> = mutableListOf()
        pcrInitData.princessMap.forEach { (t, u) ->
            val get = properties.first { it.name == characterField.field }.get(u)
            if (get == value) resultList.add(t.toInt())
        }
        if (resultList.isEmpty()) {
            return PlainText("未找到符合的人选")
        }
        val spliceCacheImage = gachaLogic.spliceCacheImage(resultList, group)
        return messageChainOf(At(sender.id), PlainText("您可能在找:"), spliceCacheImage)
    }

}
